### PR TITLE
[9.x] Migrate mail components

### PR DESCRIPTION
**Pull request by timacdonald**
This PR allows mail components to be used with the `<x-component>` syntax.

These two are now identical:

```blade
@component('mail::button', ['url' => 'https://forge.laravel.com'])
Deploy with Forge
@endcomponent

<x-mail::button url="https://forge.laravel.com">
Deploy with Forge
</x-mail>

```

Below are some of the test views I've used to ensure things are working as expected. You may like to fire these up in a standalone Mailable to test out.

```blade
<x-mail::message>

# TEST: @@component('mail::button', ['url' => ''])

@component('mail::button', ['url' => ''])
    slot content
@endcomponent

# TEST: &lt;x-mail::button url="">

<x-mail::button url="">
    slot content
</x-mail::button>

# TEST: @@component('mail::button', ['url' => 'https://forge.laravel.com'])

@component('mail::button', ['url' => 'https://forge.laravel.com'])
    slot content
@endcomponent

# TEST: &lt;x-mail::button url="https://vapor.laravel.com">

<x-mail::button url="https://vapor.laravel.com">
    slot content
</x-mail::button>

# TEST: @@component('mail::button', ['url' => 'https://forge.laravel.com', 'color' => 'success'])

@component('mail::button', ['url' => 'https://forge.laravel.com', 'color' => 'success'])
    slot content
@endcomponent

# TEST: &lt;x-mail::button url="https://vapor.laravel.com" color="success">

<x-mail::button url="https://vapor.laravel.com" color="success">
    slot content
</x-mail::button>

# TEST: @@component('mail::table')

<style> .table { background-color: palegoldenrod } </style>

@component('mail::table')
Table content
@endcomponent

# TEST: &lt;x-mail::table&gt;

<x-mail::table>
Table content
</x-mail::table>

# TEST: @@component('mail::subcopy')

@component('mail::subcopy')
## Slot content
@endcomponent

# TEST: &lt;x-mail::subcopy&gt;

<x-mail::subcopy>
## Slot content
</x-mail::subcopy>

# TEST: @@component('mail::panel')

@component('mail::panel')
## Slot content
@endcomponent

# TEST: &lt;x-mail::panel&gt;

<x-mail::panel>
## Slot content
</x-mail::panel>

# TEST: @@component('mail::header', ['url' => 'https://forge.laravel.com']) Laravel @@endcomponent

<table>
@component('mail::header', ['url' => 'https://forge.laravel.com'])
Laravel
@endcomponent
</table>

# TEST: &lt;x-mail::header url="https://vapor.laravel.com"&gt; Laravel &lt;/x-mail::header&gt;

<table>
<x-mail::header url="https://vapor.laravel.com">Laravel</x-mail::header>
</table>

# TEST: @@component('mail::header', ['url' => 'https://forge.laravel.com']) Acme @@endcomponent

<table>
@component('mail::header', ['url' => 'https://forge.laravel.com'])
Acme
@endcomponent
</table>

# TEST: &lt;x-mail::header url="https://vapor.laravel.com"&gt; Acme &lt;/x-mail::header&gt;

<table>
<x-mail::header url="https://vapor.laravel.com">Acme</x-mail::header>
</table>

# TEST: @@component('mail::footer')

<table>
@component('mail::footer', ['url' => 'https://forge.laravel.com'])
## Slot content
@endcomponent
</table>

# TEST: &lt;x-mail::footer&gt;

<table>
<x-mail::footer url="https://vapor.laravel.com">
## Slot content
</x-mail::footer>
</table>

# TEST: Pass slots instead of props...

@component('mail::button')
@slot('url')
https://forge.laravel.com
@endslot
slot content
@endcomponent

<x-mail::button>
<x-slot:url>
https://forge.laravel.com
</x-slot:url>
slot content
</x-mail::button>

</x-mail::message>

```

```blade
@component('mail::message')

# TEST: @@component('mail::button', ['url' => ''])

@component('mail::button', ['url' => ''])
    slot content
@endcomponent

# TEST: &lt;x-mail::button url="">

<x-mail::button url="">
    slot content
</x-mail::button>

# TEST: @@component('mail::button', ['url' => 'https://forge.laravel.com'])

@component('mail::button', ['url' => 'https://forge.laravel.com'])
    slot content
@endcomponent

# TEST: &lt;x-mail::button url="https://vapor.laravel.com">

<x-mail::button url="https://vapor.laravel.com">
    slot content
</x-mail::button>

# TEST: @@component('mail::button', ['url' => 'https://forge.laravel.com', 'color' => 'success'])

@component('mail::button', ['url' => 'https://forge.laravel.com', 'color' => 'success'])
    slot content
@endcomponent

# TEST: &lt;x-mail::button url="https://vapor.laravel.com" color="success">

<x-mail::button url="https://vapor.laravel.com" color="success">
    slot content
</x-mail::button>

# TEST: @@component('mail::table')

<style> .table { background-color: palegoldenrod } </style>

@component('mail::table')
Table content
@endcomponent

# TEST: &lt;x-mail::table&gt;

<x-mail::table>
Table content
</x-mail::table>

# TEST: @@component('mail::subcopy')

@component('mail::subcopy')
## Slot content
@endcomponent

# TEST: &lt;x-mail::subcopy&gt;

<x-mail::subcopy>
## Slot content
</x-mail::subcopy>

# TEST: @@component('mail::panel')

@component('mail::panel')
## Slot content
@endcomponent

# TEST: &lt;x-mail::panel&gt;

<x-mail::panel>
## Slot content
</x-mail::panel>

# TEST: @@component('mail::header', ['url' => 'https://forge.laravel.com']) Laravel @@endcomponent

<table>
@component('mail::header', ['url' => 'https://forge.laravel.com'])
Laravel
@endcomponent
</table>

# TEST: &lt;x-mail::header url="https://vapor.laravel.com"&gt; Laravel &lt;/x-mail::header&gt;

<table>
<x-mail::header url="https://vapor.laravel.com">Laravel</x-mail::header>
</table>

# TEST: @@component('mail::header', ['url' => 'https://forge.laravel.com']) Acme @@endcomponent

<table>
@component('mail::header', ['url' => 'https://forge.laravel.com'])
Acme
@endcomponent
</table>

# TEST: &lt;x-mail::header url="https://vapor.laravel.com"&gt; Acme &lt;/x-mail::header&gt;

<table>
<x-mail::header url="https://vapor.laravel.com">Acme</x-mail::header>
</table>

# TEST: @@component('mail::footer')

<table>
@component('mail::footer', ['url' => 'https://forge.laravel.com'])
## Slot content
@endcomponent
</table>

# TEST: &lt;x-mail::footer&gt;

<table>
<x-mail::footer url="https://vapor.laravel.com">
## Slot content
</x-mail::footer>
</table>

# TEST: Pass slots instead of props...

@component('mail::button')
@slot('url')
https://forge.laravel.com
@endslot
slot content
@endcomponent

<x-mail::button>
<x-slot:url>
https://forge.laravel.com
</x-slot:url>
slot content
</x-mail::button>

@endcomponent
```

### CODE REVIEW

File: src/Illuminate/View/Compilers/Concerns/CompilesComponents.php [hunk 1]
```diff
@@ -149,7 +149,8 @@ protected function compileEndComponentFirst()
      */
     protected function compileProps($expression)
     {
-        return "<?php foreach(\$attributes->onlyProps{$expression} as \$__key => \$__value) {
+        return "<?php \$attributes ??= new \\Illuminate\\View\\ComponentAttributeBag; ?>
+<?php foreach(\$attributes->onlyProps{$expression} as \$__key => \$__value) {
```
Comment (timacdonald) [inline src/Illuminate/View/Compilers/Concerns/CompilesComponents.php:153]: This ensures that the older style syntax can work with the `@props` directive.

File: src/Illuminate/View/Compilers/ComponentTagCompiler.php [hunk 2]
```diff
@@ -312,6 +312,10 @@ public function componentClass(string $component)
             return $guess;
         }
 
+        if (Str::startsWith($component, 'mail::') && $viewFactory->exists($component)) {
+            return $component;
+        }
```
Comment (timacdonald) [inline src/Illuminate/View/Compilers/ComponentTagCompiler.php:317]: I've given special treatment to `mail::` prefixed components, as they are given special treatment in the framework.

**Comment by timacdonald** [comment]
Documentation: https://github.com/laravel/docs/pull/8284

File: src/Illuminate/Mail/resources/views/text/message.blade.php [hunk 3]
```diff
@@ -1,27 +1,27 @@
...
-        @component('mail::footer')
+    <x-slot:footer>
+        <x-mail::footer>
             © {{ date('Y') }} {{ config('app.name') }}. @lang('All rights reserved.')
-        @endcomponent
-    @endslot
-@endcomponent
+        </x-mail::footer>
+    </x-slot:footer>
+</x-mail::layout>
```
Comment (timacdonald) [inline src/Illuminate/Mail/resources/views/text/message.blade.php:27]: I migrated this file, but I'll admit I don't know what this file is doing and if it even makes sense - being that it is the _text_ representation.

File: src/Illuminate/Mail/resources/views/text/message.blade.php [hunk 4]
```diff
@@ -1,27 +1,27 @@
...
```
Comment (timacdonald) [inline src/Illuminate/Mail/resources/views/text/message.blade.php:27]: I migrated this file, but I'll admit I don't know what this file is doing and if it even makes sense - being that it is the _text_ representation, but seems to be reaching for the registered HTML blade views 🤷‍♂️

**Comment by DougSisk** [comment]
I'm now getting an exception when running `view:cache`:

````
  InvalidArgumentException 

  Unable to locate a class or view for component [mail::message].

  at vendor/laravel/framework/src/Illuminate/View/Compilers/ComponentTagCompiler.php:319
````

**Pull request merged**